zookeeper-server/src/test/java/org/apache/zookeeper/test/BackupManagerTest.java

```diff
@@ -0,0 +1,601 @@
+        backedUpLog1, backupManager.getBackedupLogZxid());
+    backupManager.getLogBackup().run(1);
+    Assert.assertEquals("Backup point cannot have moved #2",
+        backedUpLog1, backupManager.getBackedupLogZxid());
+  }
+
+  @Test
+  public void testLostLogs() throws Exception {
+    createNode(connection, "/foo");
+    backupManager.getLogBackup().run(1);
+    createNode(connection, "/foobar");
+    createNode(connection, "/foobar2");
+    zks.getTxnLogFactory().rollLog();
+    createNode(connection, "/foobar3");
+    createNode(connection, "/foobar4");
+    zks.getTxnLogFactory().rollLog();
+    createNode(connection, "/foobar5");
+
+    File[] logs = snapLog.getSnapshotLogs(0);
+
+    for (int f = 0; f < logs.length; f++) {
+      LOG.info("File {}: {}", f, logs[f].getPath());
+    }
+
+    Assert.assertEquals("Should have three log files", 3, logs.length);
+    logs[0].delete();
+    logs[1].delete();
+    backupManager.getLogBackup().run(1);
+    backupManager.getLogBackup().run(1);
+    backupManager.getLogBackup().run(1);
+
+    String[] backupFiles = getBackupFiles(backupDir);
+    Assert.assertEquals("Should have three files in backup dir", 3, backupFiles.length);
+
+    boolean lostLogsFileFound = false;
+    for (String s : backupFiles) {
+      LOG.info("backupFile: {}", s);
+      if (s.equals("lostLogs.3-6")) {
+        lostLogsFileFound = true;
+        break;
+      }
+    }
+
+    Assert.assertTrue("Must have a lostLogs.3-6 file", lostLogsFileFound);
+  }
+
+  @Test
+  public void testCleanupOfInvalidFilesOnStartUp() throws Exception {
+    createNodes(connection, 110);
+    backupManager.getSnapBackup().run(1);
+    backupManager.getLogBackup().run(1);
+    createNodes(connection, 110);
+    backupManager.getSnapBackup().run(1);
+    backupManager.getLogBackup().run(1);
+    createNodes(connection, 110);
+    backupManager.getSnapBackup().run(1);
+    backupManager.getLogBackup().run(1);
+    createNodes(connection, 110);
+    backupManager.getSnapBackup().run(1);
+    backupManager.getLogBackup().run(1);
+
+    String[] expectedBackupFiles = getBackupFiles(backupDir);
+
+    backupManager.initialize();
+
+    String[] actualBackupFiles = getBackupFiles(backupDir);
+
+    Assert.assertArrayEquals("no files should have been deleted",
+        expectedBackupFiles, actualBackupFiles);
+
+    // Add some invalid files
+    File badFile = new File(
+        backupDir,
+        makeTmpName(BackupUtil.makeBackupName(Util.makeSnapshotName(1000), 1010)));
+    badFile.createNewFile();
+
+    badFile = new File(
+        backupDir,
+        makeTmpName(BackupUtil.makeBackupName(Util.makeSnapshotName(1100), 1110)));
+    badFile.createNewFile();
+
+    badFile = new File(
+        backupDir,
+        makeTmpName(BackupUtil.makeBackupName(Util.makeLogName(1000), 1200)));
+    badFile.createNewFile();
+
+    backupManager.initialize();
+
+    actualBackupFiles = getBackupFiles(backupDir);
```

Comment:
You may want to modify the getBackupFiles method and add name.startsWith("TMP_"). But please be aware that logs and snapshots are stored under different directories. See my comment on getBackupFiles().

Reply:
@mgao0 Upon giving this a second thought, do you think we should store them all in the same directory and not make the subfolders like /log or /snapshot? How hard do you think it would be to make this change?

This way we can keep this test structure (same as the Twitter design).

Unless you have a strong reason not to. Feel free to add your thoughts here.

Reply:
Please see my reply above in the other comment. Resolving this one.